tools/validate/validate.cpp

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#define _CRT_SECURE_NO_WARNINGS
+#include <algorithm>
+#include <array>
+#include <assert.h>
+#include <exception>
+#include <filesystem>
+#include <stdexcept>
+#include <stdio.h>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+using namespace std;
+
+class BinaryFile {
+public:
+    explicit BinaryFile(const string& filepath) {
+        m_file = fopen(filepath.c_str(), "rb");
+
+        if (!m_file) {
+            throw runtime_error("fopen() failed for: " + filepath);
+        }
+    }
+
+    [[nodiscard]] bool read_next_block(vector<unsigned char>& buffer) {

```

Comment:
Why `unsigned char` instead of `char8_t`?

Reply:
I am intentionally reading raw bytes and performing no interpretation (not even line endings), for the purpose of rejecting characters outside of the printable ASCII range (plus CR, LF, and sometimes tabs). Using `char8_t` would indicate that UTF-8 interpretation is the default, so I believe that it would be misleading, in addition to not providing any type safety or other benefits.

For somewhat related reasons, I'm not using `byte`. I could, because it's an opaque-ish byte type, but in at least one case I want to perform a value range comparison.

Reply:
That's reasonable. I'll allow that `unsigned char` - or perhaps even `uint8_t` - is the right type here.